source/Halibut/Transport/AuthorizedTcpConnectionsLimiter.cs

```diff
@@ -0,0 +1,86 @@
+using Halibut.Diagnostics;
+using Halibut.Exceptions;
+
+namespace Halibut.Transport
+{
+    public interface IAuthorizedTcpConnectionsLimiter
+    {
+        IDisposable ClaimAuthorizedTcpConnection(string thumbprint);
+    }
+
+    public class AuthorizedTcpConnectionsLimiter : IAuthorizedTcpConnectionsLimiter
+    {
+        readonly HalibutTimeoutsAndLimits timeoutsAndLimits;
+
+        //we use a StrongBox here so we can use Interlock to increment/decrement the value
+        ConcurrentDictionary<string, StrongBox<int>> authorizedConnectionCountPerThumbprint = new();
+
+        public AuthorizedTcpConnectionsLimiter(HalibutTimeoutsAndLimits timeoutsAndLimits)
+        {
+            this.timeoutsAndLimits = timeoutsAndLimits;
+        }
+
+        public IDisposable ClaimAuthorizedTcpConnection(string thumbprint)
+        {
+            return new AuthorizedTcpConnectionLease(thumbprint, authorizedConnectionCountPerThumbprint, timeoutsAndLimits.MaximumAuthorisedTcpConnectionsPerThumbprint);
+        }
+
+        class AuthorizedTcpConnectionLease : IDisposable
+        {
+            readonly string thumbprint;
+            readonly ConcurrentDictionary<string, StrongBox<int>> authorizedConnectionCountPerThumbprint;
+
+            public AuthorizedTcpConnectionLease(string thumbprint, ConcurrentDictionary<string, StrongBox<int>> authorizedConnectionCountPerThumbprint, int maximumAcceptedTcpConnectionsPerThumbprint)
+            {
+                this.thumbprint = thumbprint;
+                this.authorizedConnectionCountPerThumbprint = authorizedConnectionCountPerThumbprint;
+
+                var count = this.authorizedConnectionCountPerThumbprint.GetOrAdd(thumbprint, _ => new StrongBox<int>(0));
+
+                //increment the count via an interlock
+                var resultCount = Interlocked.Increment(ref count.Value);
+
+                //validate the new count. If this throws an exception, it'll
+                if (resultCount > maximumAcceptedTcpConnectionsPerThumbprint)
+                {
+                    //decrement as this connection has been rejected
+                    Interlocked.Decrement(ref count.Value);
+                    throw new AuthorizedTcpConnectionsExceededException($"Exceeded the maximum number ({maximumAcceptedTcpConnectionsPerThumbprint}) of authorised TCP connections for thumbprint {thumbprint}");
+                }
+            }
+
+            public void Dispose()
+            {
+                if (authorizedConnectionCountPerThumbprint.TryGetValue(thumbprint, out var count))
+                {
+                    //decrement the count of authorized connections
+                    var result = Interlocked.Decrement(ref count.Value);
+
+                    //if this was the last connection, remove the value from the dictionary
+                    if (result == 0)
+                    {
+                        authorizedConnectionCountPerThumbprint.TryRemove(thumbprint, out _);
```

Comment:
This will likely need to lock the entire concurrent dict for this to be safe. Since something else could come in increment the count while this removes it.

Reply:
yeah, it's not perfect. Working on it

Reply:
Yeah, consider whether we could remove a lot of this complexity by simply using a regular Dictionary and one lock for retrieval, increment and comparison in the contractor and one lock in the Dispose method.   We are not doing any IO here so this code should lighting fast.

Reply:
I've come up with a good solution that uses the concurrent dictionary with no extra locking

Reply:
There seem to be at least 2 more problems in the constructor.

1. The factory method passed to GetOrAdd can be executed [multiple times](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getoradd?view=net-8.0) which means that this code can allow more connections than what is specified by the limit. E.g. 3 threads call `GetOrAdd` with the same thumbprint and each creates a new instance of  `ConnectionCount` but only one of them will be stored in the dictionary and the value of count will be 1.
2. Interlocked.Increment and check are not synchronized which means that legitimate connections can be rejected. E.g. 3 threads with the same thumbprint increase the the counter before they move to the check on line 63.  Now, if the limit is 2 then all of them will be rejected.

Multithread code is hard, very hard to write and even harder to maintain for the future us.   Please reconsider the approach for something simpler. I don't think this code has to be optimized to the nth degree.

Reply:
This might be worth discussing on a call because my understanding and testing of GetOrAdd and Interlocked should not behave the way you are suggesting in both scenarios

Reply:
Ok, I will schedule something today.

Reply:
Thanks @APErebus for your time. I got both of those scenarios wrong and learnt something new about the way these synchronization primitives work.  

That said, @APErebus had to consult Stephen Toub's work to implement this solution, which is a good indicator re how easy/hard it will be to maintain.  We agreed that the increased level of complexity might not be warranted for this solution and @APErebus will switch to a regular lock.